### PR TITLE
Proof-of-concept for "snow ws" command generation

### PR DESCRIPTION
**Pull request by sfc-gh-cgorrie**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.

### Changes description

Generates `snow ws @<entity_id>` command groups based on the current directory's `snowflake.yml`.

What's done:
* allows decoupling commands from classes, so that each one can live in its own compilation unit
* defines an annotation DSL (e.g. `HelpText`, `ParameterDeclarations`) that ensures individual commands don't actually rely on click / typer
* generates click / typer commands from those annotations
* allows commands like "version create" to automatically create sub-trees

Things that need addressing:
* deal with position action arguments (e.g. `snow @pkg version create v1` needs to be passed in instead as `--version v1`)
* tests are failing
* commands cannot share option arguments (they all need to re-define them)

### CODE REVIEW

File: src/snowflake/cli/_plugins/workspace/entity_commands.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    for each found entity. Throws a fatal ClickException if templating is used
+    in the basic definition of entities (i.e. the type discriminator field).
+    """
+    dm = DefinitionManager(str(project_root) if project_root is not None else None)
+
+    if not dm.has_definition_file:
+        return
+
+    root = dm.unrendered_project_definition
+    for (target_id, model) in root.entities.items():
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/_plugins/workspace/entity_commands.py:94]: Should we consider `model.get_command_group(target_id)` that would encapsulate L100-102? WDYT?

Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/_plugins/workspace/entity_commands.py:180]: I'm not sure I quite understand the suggestion. Can you give a bit more detail?

File: src/snowflake/cli/_plugins/workspace/entity_commands.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            ws = WorkspaceManager(
+                project_definition=cli_context.project_definition,
+                project_root=cli_context.project_root,
+            )
+
+            # filter out common CLI options our action doesn't take
+            pass_through_options = {
+                k: v for (k, v) in options.items() if k in known_params
+            }
+            ws.perform_action(self.target_id, action, **pass_through_options)
```
Comment (sfc-gh-fcampbell) [inline src/snowflake/cli/_plugins/workspace/entity_commands.py:141]: how do we handle return values? for example, we currently have some command that wrap the return value of the entity action in `MessageResult`, others use `QueryResult`, etc

File: src/snowflake/cli/_plugins/workspace/entity_commands.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            return MessageResult(
+                f"Successfully performed {action.verb} on {self.target_id}."
+            )
```
Comment (sfc-gh-fcampbell) [inline src/snowflake/cli/_plugins/workspace/entity_commands.py:144]: not all actions want to show a success message (especially those that retrieve data, like `version-list`)

**Comment by sfc-gh-fcampbell** [comment]
Closing this for now due to uncertainty regarding the `snow ws` command. We can use the ideas from this PR in a future re-attempt if necessary.

**Pull request closed without merging**